externallib/DPA/DP.py

```diff
@@ -0,0 +1,152 @@
+"""
+Module implementing the "Fast Search and Find of Density Peaks" clustering algorithm of Rodriguez and Laio (2014).
+"""
+
+import numpy as np
+import scipy as sp
+from scipy import stats
+from sklearn.neighbors import NearestNeighbors
+
+
+class Density_Peaks_clustering:
+    """
+    Clustering by fast search and find of density peaks, Rodriguez and Laio (2014).
+    """
+    def __init__(self, distances=None, indices=None, dens_type="eps", dc=None, percent=2.0):
+
+        """
+
+        Parameters
+        ----------
+        distances: Matrix of distances between data points and their neighbours of dimension Nxn_neigh_comp where N is
+        the number of data points and n_neigh_comp is the number of neighbours to consider.
+        indices: The indices in the data matrix of the neighours for all of the data points.
+        dens_type: The type of density to compute (can be exponential)
+        dc: The cutoff distance beyond which data points don't contribute to the local density computation of another
+        datapoint.
+        percent: Criterion for choosing dc, int from 0-100, typically chosen such that the average number of neighbours is 1-2% of the
+        total number of points in the dataset.
+        """
+
+        self.distances = distances
+        self.indices = indices
+        self.dens_type = dens_type
+        self.dc = dc
+        self.percent = percent
+        self.dens = None  # numpy array of the densities of the data points
+        self.delta = None  # numpy array of the distances to the nearest cluster centre
+        self.ref = None  # numpy array of the indices of the cluster centres for each data point
+        self.decision_graph = None
+        self.delta_cut = None
+        self.dens_cut = None
+        self.cluster = None
+        self.centers = None
+        self.halo = None
+
+    def get_dc(self, data):
+
+        """
+        Compute the cutoff distance given the data.
+
+        Parameters
+        ----------
+        data: The dataset of dimension mxn where m is the number of data points and n is the number of features.
+
+        Returns
+        -------
+
+        self.dc, the cutoff distance
+
+        """
+
+        Nele = data.shape[0]  # The number of data points
+        n_neigh = int(self.percent/100.*Nele)
+        n_neigh_comp = int(4.*self.percent/100.*Nele)
```

Comment:
If we are to include this class it would be good to explain what this **n_neigh_comp** variable is as it took me a long time to figure out.

Reply:
Do you know what this is?

Reply:
Do you mean where is this n_neigh_comp in the original Laio_DB class that I wrote?

Reply:
The physical meaning